### ISSUE TITLE
Allow registered users to update their ballot

### ISSUE THREAD

**Issue by rupalkhilari** [electology/approval_frame#75]
If a poll has been marked as type 2 (only registered users can vote), it would be nice to allow the user to update his/her ballot anytime before the closing date of the poll is reached (to be implemented in https://github.com/electology/approval_frame/issues/15) . Currently, with this setting, the user is only allowed to enter their ballot once, after which he/she is not allowed to vote in that poll again.

**Comment by kushankr** [comment]
I am going to work on this. Also, related to this, I will allow the user to select 'poll closing date' on 'Create' page. This closing date can be later edited using 'Edit' functionality (another ticket).

**Comment by kushankr** [comment]
@rupalkhilari 

This is going to be a huge post, Sorry. This post also related to #81 
I wanted to request feedback on the following:

1) As I mentioned that I am going to allow user to select a closing date in 'create' page, right now i am currently placing it in 'poll settings' panel in 'create' page. The question is related to TIME ZONE. The poll closing date should be in the timezone of the user right? so if i am in MA, it should show date corresponding to us/eastern or if u r in SF, it should show PST. or the dates should be all UTC? coz depending on time, there will be difference in date of a UTC vs PST vs EST. Is the date always stored in UTC in the DB?
If you can share your thoughts on this, it would be great. 

2) Is selecting a closing date compulsory? if the user does not select one, What should happen?  Right now, i am putting current date as default in closing date field in 'create' page.  How about allowing user to select a closing date only starting from 1 day after the day poll was created (so that a poll is open for a minimum of one day)?

3) Before closing date, a user can update his/her vote. After closing date, i was thinking that I would redirect to results page and a message that poll is closed?

4) 'Edit a poll' issue - sparky was working on it and i saw in his code that he has modified models in his code branch to include close_date and I was wondering if i can just copy his code around close_date and put it in my branch. If he comes back to finish the functionality, i would not want him to refactor his code based on my model.

I would like to share SCs of my work 
![issue 75 - 1](https://cloud.githubusercontent.com/assets/8802245/11887366/8304b42c-a501-11e5-9ed3-7b346d4505f5.JPG)
![issue 75 - 2](https://cloud.githubusercontent.com/assets/8802245/11887367/866ec7ce-a501-11e5-85af-97987a69d61f.JPG)

**Comment by rupalkhilari** [comment]
Hi Kushank,

Thanks so much for looking into these related issues! I have commented in the order you mentioned. 

1) Yes, all dates and times should be in the timezone of the user in order to avoid confusion(since the user will be selecting a date keeping that in mind). Currently, since we have set USE_TZ=True in settings.py, all datetime objects are converted into  UTC time and stored into the database.
You perhaps might have checked https://docs.djangoproject.com/en/1.8/topics/i18n/timezones/  wherein they have advised the use of the 'localtime' template filter (in the .html templates) and the django.utils.timezone.localtime (in .py files) to print timezone aware datetimes from regular UTC datetimes in the database. 

2) I believe selecting a closing date should not be made compulsory. The user should be allowed to choose to keep the poll open forever as well (perhaps as the default).

3) After the poll is closed, instead, I think that the vote button should be disabled, with a small message above/below the button which states something on the lines of 'Sorry! This poll has been closed.' In this case the user could choose the 'See Results' link in case he/she wishes to see them. - Perhaps in the future we might want the 'view results' also to be configured - so that the results are only viewable after a certain date and not during the voting period. 

4) Thats a great idea, if that conforms to what you also have in mind it would be nice to reuse it. 

The integration into the Poll Settings panel and the Date widget looks great from the screenshots. Just a couple of suggestions.
- Since we will be using timezone aware datetimes, the user should also be allowed to select the time of the day, and it should reflect accordingly across timezones.
- The label could read 'Please select a closing date and time for this poll, if any', allowing the date/time fields to be cleared as well.

Please let me know in case you have any concerns, suggestions or in case I missed something.

Thanks!

**Comment by kushankr** [comment]
Thanks Rupal!

I am fine with everything but regarding the screenshots, what I initially planned is that I will let the user to select only a date. My idea was as soon as a new date arrives, poll is closed. So the current calendar is showing date according to timezone of the user. I also felt that this is what Sparky was coding towards in his models code. 

But anyways, I will try to create a way for user to pick a time too. If I cannot get a calendar view working for time, I will allow user to select times from a drop down. I should let you know that I am using third party bootstrap js and css for datepickers.  And for future, I will also be using either jquery datetime picker or bootstrap datetimepicker (both require downloading a .js and .css file to vendor folder).

**Comment by kushankr** [comment]
Sorry to bother you again, Rupal. I just wanted confirmation on one more item: If the user has already voted, instead of a 'Vote' button, i want to show 'Update Vote' button and a message saying that the user has already voted in the poll and can update the vote before 'closing_date'. One more thing: A user can update vote any number of times before closing date right? 

Thank you for your valuable feedback, Rupal. I am asking all this now because from next week, holiday season might start and I do not want to keep you busy here. I will work on this code and submit a PR in the first week of January, 2016.

**Comment by rupalkhilari** [comment]
Hi Kushank, 

Yes, I see your point about having the date alone, but I wonder if that would lead to ambiguity in different timezones. Not sure if i am missing something though. Eg. The poll closing date is set to 25th Dec. 2015 IST, so if the closing date for display in PST is 24th Dec. 2015, it would allow lesser hours for the voter there, and if it showed 25th Dec. 2015, it would allow a few more hours for the voter to vote (compared to IST). Thus could it happen that the poll owner in India (IST) may believe that the poll is closed, when someone in PST can still vote in it ? But you have mentioned you are looking into  the time as well, so I'm sure you will take care of it. Also using third party bootstrap/jquery widgets should be good as long as they are well supported (although first preference is always to native controls).

Regarding your question about 'Update Vote', I too have the same thoughts about its functionality. For this, along with displaying 'Update vote' perhaps you will also be displaying the current ballot selections. Perhaps later on we can look into making the number of times the vote can be updated configurable, if required by Aaron. 

Thanks for taking the time out to work on this project and bring in these very essential features. Since they are quite major, please do take your time - you are more than welcome to contribute anytime :smiley:

**Comment by kushankr** [comment]
Thank you Rupal for your feedback.

Regarding the timezone issue, this is what I have in mind now:

 I will explain by an example.

So say I am a user in MA and I selected 20th Dec, 2015 10:45 A.M. as closing date and time. You are in PST and you will see closing date and time as the one I selected and with EST added as timezone/ see corresponding closing date and time in PST. For you poll will close at 20th Dec, 7:45 A.M. PST in this scenario. 

If say i created poll at 19th Dec, 10:45 A.M. , i will have 24 hours to complete the poll. For you in PST, the poll creation time will be 19th Dec, 7:45 A.M. (since you are 3 hours behind) and since the poll for you will close at 20th Dec 7:45 A.M., you will have 24 hours too. It will never occur that one person in one timezone is able to vote while poll is closed for other user because at 20th Dec 10:45 A.M. EST, poll for me will close and it will be 20th Dec 7:45 A.M. PST for you and you will see poll closed too.

If I were to say that 20th Dec 10:45 A.M. should be the closing date and time for all users irrespective of time zone (so 20th DEC 10:45 A.M. EST for me and 20th DEC 10:45 P.S.T for you too), then you in PST will get more hours to vote. Because: for me poll is open between 19th Dec 10:45 A.M. EST and 20th Dec 10:45 A.M. EST but for you it is open 19th Dec 7:45 A.M. PST - 20th Dec 10:45 A.M. PST. 

Is this what you were getting at too? I thought that I will explain my understanding of how datetime will operate so that if i am missing something, you can correct me.

**Comment by rupalkhilari** [comment]
Thanks for the detailed explanation, @kushankr . Yes, I thought of it in the same way too - including both date and time. Timezone is indeed always an important consideration - thanks for identifying it and taking care of it.

**Comment by kushankr** [comment]
I have been dealing with the timezone issue

Here is the line in settings.py file

# Local time zone for this installation. Choices can be found here:
# http://en.wikipedia.org/wiki/List_of_tz_zones_by_name
# although not all choices may be available on all operating systems.
# In a Windows environment this must be set to your system time zone.
TIME_ZONE = 'America/Los_Angeles'

That is why when I create a poll, the time shown in pub_date below a poll is LA time. 

When storing closing date in DB, it always takes LA time as ref. So if i am saving 22 dec 2015, 4:30 a.m. , UTC should be 22 dec 2015, 9:30 a.m. but it shows 22 dec 2015, 12:30 p.m. in DB because it is considering 22 dec 2015, 4:30 a.m. as LA time.

I have been reading about this and it is said that ideally TIME_ZONE should be UTC in settings.py file. What do you think, @rupalkhilari ?

**Comment by kushankr** [comment]
Ok, so I just found a way to save user's local date time in UTC DB without it using LA timezone. I had to factor in the offset and set the tzinfo. It is always weird: I spend time figuring out something for hours and i message you and immediately after that, I come up with a way to fix it!

However, I was wondering @rupalkhilari  what do you think of changing TIME_ZONE = 'America/Los_Angeles' to TIME_ZONE = 'UTC' in settings.py , I want 'poll published date' to show local time of user and not corresponding LA Time? Also, it is good to set it to UTC and display date time depending on local timezone of user.

**Comment by kushankr** [comment]
I have deleted my previous messages to avoid any confusion.

My problem was: I was not able to find a way to save user's local time in DB. As we know, DB stores time in UTC. However, it requires timezone info. I was not supplying the timezone info and it was using default time zone (LA time) from settings.py. I found a way to save user's local date time in UTC DB without it using LA timezone. I had to factor in the offset and set the time zone. But I saw something that I have been noticing for a while. 

If you see polls page and you see published date, it shows LA time. This is true for test application server too. As an example, i created a poll at 6:30 a.m. eastern time, however it was showing that poll is created on 22 dec, 3:30 a.m. (LA Time). I think that it is because of TIME_ZONE = 'America/Los_Angeles' in settings.py. Now, I also saw that publishing date is storing in the database correctly because it uses timezone.now() and this is not affected by TIME_ZONE settings. However, when the date is displayed, it always displays the LA time. I think that this should not happen. If i am a user, I would not like to see poll creation date as anything but corresponding to my local time.

I have a suggestion: What do you think of changing TIME_ZONE = 'America/Los_Angeles' to TIME_ZONE = 'UTC' in settings.py , it is good to set it to UTC and display date time depending on local timezone of user. I can look into it but i thought i will ask you if this is a BUG that poll creation date is not corresponding to local timezone of user? I can still work on my issues without changing it in settings.py but since we are on this issue of timezones, i thought that it is better to resolve it now.

**Comment by rupalkhilari** [comment]
Yes, it is a bug, it should have ideally displayed the user's local time. I believe I didn't notice that since I am also in the 'America/Los_Angeles' timezone and hence the datetime beside the published polls appears correct to me. You should go ahead and change TIME_ZONE to UTC in the settings.py for consistency. All datetimes in the database are in UTC. 

You mentioned that you are using offsets to take care of the timezone, is it using Javascript or pytz for finding the local timezone in the template ? There also seem to be ways to accomplish this with filters and [middleware](https://docs.djangoproject.com/en/dev/topics/i18n/timezones/#selecting-the-current-time-zone). Might need to evaluate these options.

**Comment by kushankr** [comment]
Thank you, Rupal

I will change it to UTC.

Regarding timezone, i am using jquery datepicker now. Because third party bootstrap datepickers have a few bugs. So, in the.js file, I am using a javascript offset function (one line) to get offset and send it to form. 

I evaluated other options too including pytz or middleware - 
i did not want user to select timezones from a dropdown in template. I wanted to auto-detect it. My js code (and it does this in one line) is able to do that. And it takes care of day light savings time too so that in future, no bug arises. Also, you need to set the timezone info of user and the only way for me to do that was to send the offset. 

I explored a lot of options and looked at Django documentation and pytz but I learned that there are only 2 options 1) send offset using js  2) allow user to select timezones and get timezone info from that. I personally prefer 1 but what do you think? 

Now when i change time_zone to utc in settings.py, i might not need to send offset but i will look into it.

**Comment by rupalkhilari** [comment]
It is good that official jQuery datetime picker would serve your purpose - always preferred to third party controls. 
As you have mentioned, I would also not prefer having the user select his/her timezone - it seems like an unnecessary overhead for the user to manage. Centralized javascript code for this should be totally fine. 
Thanks!

**Comment by kushankr** [comment]
@rupalkhilari 

So, I wanted to ask you if i can install a 3rd party python package named 'tzlocal' to fix our timezone issues. I have researched everything and tried everything - to fix published date bug, i have to install 'tzlocal' and use django middleware to set timezone of app to timezone of user. 

I was using javascript offset to help me with my close_date (it only sends timezone offset) but i think that a more elegant solution is to use tzlocal to get current timezone of user and use django middleware. This will help us in future too with all timezone issues. It is also more consistent with django framework. Please let me know.

**Comment by kushankr** [comment]
@rupalkhilari 

I have found a way using 'pytz' module instead of 'tzlocal'. I think it is fine to use pytz and django middleware to resolve our timezone issues right?

**Comment by rupalkhilari** [comment]
Yes @kushankr , the middleware solution with pytz seems to be a recommended one, though I was wondering if tzlocal would return the local timezone of the server rather than the client. I believe the database entries would retain the UTC timezone after the timezone.activate(). 

Since you are not making using of Javascript at the client, could you outline your method of finding the local timezone to be set using timezone.activate() in the middleware ? (it this using only pytz, both or other dateutils ?)

**Comment by kushankr** [comment]
Hi Rupal,

Thank you for your reply.

 1) tzlocal method is get_localzone() -> it returns the local timezone of client and not the server.

2) for pytz -> i am using reference.LocalTimezone()

I am not using both - we can use either 1 or 2. tzlocal which is a part of dateutils is recommended in most places. In pytz docs, it is said that pytz.reference should be used only for tests [pytz - reference.LocalTimezone()] 

I would prefer to use tzlocal, I included pytz because it was serving our purpose.

**Comment by kushankr** [comment]
I will research more on this, Rupal. See if i can get more options to use for timezone.activate(). 
However, i thought that i would ask you: Do you have any suggestions on this?

**Comment by rupalkhilari** [comment]
Yes @kushankr, can you elaborate at what stage you are attempting to run reference.LocalTimezone()/ tzlocal.get_localzone() ?
. I did a few tests with setting TIME_ZONE to UTC in settings.py, and the server seemed to be returning 'UTC' itself instead of the local timezone. Let me know in case I am missing something.
I think this indeed is a tricky problem to solve, I was wondering if we might need to detect the user's timezone name ('America/Los_Angeles') via javascript using something like [jstz](http://pellepim.bitbucket.org/jstz/). 
Once that is done, there could be a couple of possibilities:
1) It could be saved into the request object and activated in the middleware. (more consistent solution)
2) The output could be modified at the client side in the template using the ['timezone' template tag](https://docs.djangoproject.com/en/1.8/topics/i18n/timezones/#template-tags)

**Comment by kushankr** [comment]
I am using reference.LocalTimezone()/ tzlocal.get_localzone()  in a custom middleware class. So i created a folder named middleware in approval_frame/approval_frame and created a class in it. Then i registered this middleware class in settings.py. 

My code of middleware class:
from django.utils import timezone
from tzlocal import get_localzone
from pytz import reference

class TimezoneMiddleware(object):
    def process_request(self, request):
        tzname = get_localzone()
        if tzname:
            timezone.activate(tzname)
        else:
            timezone.deactivate()

Oh, you need to remove TIME_ZONE in settings.py because now we are using middleware to set it. Also, i think that template uses default TIME_ZONE specified in settings.py to render date time. The date time is always stored in UTC in DB regardless, the TIME_ZONE setting is for developer to set the timezone of the app and since in our case it is dynamically set, we should remove it.

'jstz' is something I came about. I decided to search for alternatives as the project still has 15 open issues. And i was not sure if activate() will take the timezone names but I will look into it.

I was just wondering: where to put 'var tz = jstz.determine();' ?. I want user's timezone to be activated when user sees polls in index page or my_polls page. If i put it in 'create.js' and process it, it will only activate() when a user creates a poll. 

Yes, this is a tricky issue. I am going to try to work with 'jstz' and will update you.

**Comment by kushankr** [comment]
@rupalkhilari 

Can you plz elaborate more on this?
* Once that is done, there could be a couple of possibilities:
it could be saved into the request object and activated in the middleware. (more consistent solution)*

I am planning to put 'var tz = jstz.determine();' in a js file for index.html or a .js file for base.html -> append it as a hidden input -> and send it as a part of request object -> set our django timezone in views.py using  "request.session['django_timezone'] = request.POST['timezone']".  Something on lines of this.

**Comment by rupalkhilari** [comment]
@kushankr , sorry for the delayed response. But using tzlocal in the middleware (in process_request) would return the local timezone of the server and not the client. I did a test on the server to verify this. Sadly, I haven't been able to find a way other than Javascript to reliably find out the client's timezone name, and jstz seemed to be doing well enough for our requirements.

By the point I had suggested, I meant exactly the same as you interpreted it next. I believe you could do this computation in the base.html instead of index.html since this code might need to get invoked even when a single poll is accessed. I shall raise a separate issue concerning this since it in itself is a major design/user experience decision we have to make.

**Comment by rupalkhilari** [comment]
I have raised https://github.com/electology/approval_frame/issues/88 to track this timezone issue (although most of the discussion has happened here). We could use this current issue to track the 'updating ballot' requirement. Thanks!

**Comment by kushankr** [comment]
@rupalkhilari 

Thank you, Rupal! Thank you for suggesting 'jstz'. I also think that using javascript to find out client's timezone is an optimal solution. 

Thank you for raising this as a separate issue. I am working on it.

**Comment by kushankr** [comment]
Just wanted to update: I am working on this issue after #88

**Comment by kushankr** [comment]
Initiated a PR
#91 

Please review it, Rupal and let me know what you think.

**Comment by rupalkhilari** [comment]
Implemented in https://github.com/electology/approval_frame/pull/91 
Thanks @kushankr !

**Comment by rupalkhilari** [comment]
Hi Kushank, 

I'm so sorry, but i might need to reopen this issue for a couple of issues I noticed:
1) If the page is not refreshed, I am able to vote/update a vote after the poll's closing datetime. (eg. Created a poll with closing datetime as today 1.00pm, if my page is open before the closing time, the button is still activated after the closing time and hence i can vote/update vote.)
2) In the poll settings panel, if I click on the date alone in the datetime picker, it chooses the current time instead of the current rounded time i.e. 00 or 30 minutes) 

Do let me know in case of any concerns. 
Thanks!

**Comment by kushankr** [comment]
Oh, the first issue - I will resolve it.

Reg the second issue, the current time is selected because it is an inbuilt functionality. Anyways, I will try to fix it too.

**Comment by rupalkhilari** [comment]
Great ! Thanks, @kushankr

**Comment by kushankr** [comment]
Initiated a PR.
#92 

Thank you, Rupal for pointing out the issues. I have thoroughly checked the functionality of jquery datetimepicker this time. It seems to be all good. But please let me know if you think I missed anything.

**Comment by rupalkhilari** [comment]
Thanks, Kushank ! Everything looks perfect  :smiley: 
Closing this https://github.com/electology/approval_frame/pull/92
